File: elixir/day11_part2_2015.exs
defmodule Day11 do
  def call() do
    initial_password = File.read!("input.txt") |> String.trim()
    first_new_password = find_next_valid_password(initial_password)
    # Start from the next password after the first new password
    second_new_password = first_new_password |> increment_password() |> find_next_valid_password()
    IO.puts(second_new_password)
  end

  defp find_next_valid_password(password) do
    Stream.iterate(password, &increment_password/1)
    |> Enum.find(&valid_password?/1)
  end

  defp increment_password(password) do
    password
    |> String.to_charlist()
    |> Enum.reverse()
    |> increment_charlist()
    |> Enum.reverse()
    |> List.to_string()
  end

  defp increment_charlist([122 | t]), do: [97 | increment_charlist(t)] # 'z' to 'a', carry the increment
  defp increment_charlist([h | t]) when h < 122, do: [h + 1 | t] # Increment char
  defp increment_charlist([]), do: ['a'] # Handle empty, start with 'a'

  defp valid_password?(password) do
    has_straight?(password) &&
      !Regex.match?(~r/[iol]/, password) &&
      has_two_pairs?(password)
  end

  defp has_straight?(password) do
    password
    |> String.to_charlist()
    |> Enum.chunk_every(3, 1, :discard)
    |> Enum.any?(fn [a, b, c] -> b - a == 1 && c - b == 1 end)
  end

  defp has_two_pairs?(password) do
    pairs = Regex.scan(~r/(.)\1/, password) |> Enum.uniq()
    length(pairs) >= 2
  end
end

Day11.call()

File: matlab/day11_part2_2015.m

function main()
    % Read the initial password from input.txt
    try
        fid = fopen('input.txt', 'r');
        if fid == -1
            error('Could not open input.txt for reading.');
        end
        current_password = strtrim(fscanf(fid, '%s'));
        fclose(fid);
    catch ME
        fprintf(2, 'Error reading input file: %s\n', ME.message);
        return;
    end

    % Find the next valid password
    next_password = find_next_valid_password(current_password);

    % Print the output to standard output
    fprintf('%s\n', next_password);

    % For Part Two, find the password after the first one
    next_password_part2 = find_next_valid_password(next_password);
    fprintf('%s\n', next_password_part2);
end

function is_valid = is_password_valid(password)
    % Rule 1: Must contain at least one increasing straight of three letters
    has_straight = false;
    for i = 1:(length(password) - 2)
        if password(i) + 1 == password(i+1) && password(i+1) + 1 == password(i+2)
            has_straight = true;
            break;
        end
    end

    % Rule 2: Cannot contain the letters i, o, or l
    has_forbidden_chars = any(password == 'i') || any(password == 'o') || any(password == 'l');

    % Rule 3: Must contain at least two different, non-overlapping pairs of letters
    pairs = [];
    i = 1;
    while i <= length(password) - 1
        if password(i) == password(i+1)
            pairs = [pairs, password(i)];
            i = i + 2; % Skip the next character as it's part of the pair
        else
            i = i + 1;
        end
    end
    has_two_pairs = length(unique(pairs)) >= 2;

    is_valid = has_straight && ~has_forbidden_chars && has_two_pairs;
end

function incremented_password = increment_password(password)
    password_chars = char(password);
    n = length(password_chars);
    i = n;
    while i >= 1
        if password_chars(i) == 'z'
            password_chars(i) = 'a';
            i = i - 1;
        else
            password_chars(i) = password_chars(i) + 1;
            break;
        end
    end
    incremented_password = char(password_chars);
end

function next_valid = find_next_valid_password(start_password)
    current = start_password;
    while true
        current = increment_password(current);
        if is_password_valid(current)
            next_valid = current;
            break;
        end
    end
end

% Call the main function to start the program
main();
